### PR TITLE
Issue#56: Implement exception handling

### PR DESCRIPTION
**Pull request by isvyry**
## ZenHab

* [Main ZenHab ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009/issues/ita-social-projects/home/56)

## Code reviewers

- [ ] @alexorlenko
- [ ] @Arthurpanda  
- [ ] @mykytam 
- [ ] @omelchenkotimur  
- [ ] @reeztl 

### Second Level Review

- [ ] @DzigMS 
- [ ] @kh0ma 

## Summary of issue

Implement initial exception handling in our application

## Summary of change

An error message is sent with exception details when it is caught

## Testing approach

---

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

**Review by alexorlenko — APPROVED** [review]

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/AlreadyExistExceptionMapper.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * AlreadyExistExceptionMapper class is used to return exception
+ * response with HTTP status code, when the element already exists.
+ *
+ * @see javax.ws.rs.ext.ExceptionMapper
+ *
+ * @author Ihor Svyrydenko
+ */
+@Provider
+public class AlreadyExistExceptionMapper implements ExceptionMapper<AlreadyExistException> {
```
Comment (kh0ma) [inline home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/AlreadyExistExceptionMapper.java]: Please introduce abstract BaseApiException. Extend `AlreadyExistException` and `NotFoundException` from it. Also pay attention to the naming of the exception, in this case, would be better to name it `NotFoundApiException`.

From the mapper's perspective please handle the mapping to the response in the abstract BaseApiExceptionMapper. Move the changeable data to the abstract methods.

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/AlreadyExistExceptionMapper.java (hunk 1 above)
Comment (mykytam) [reply home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/AlreadyExistExceptionMapper.java]: Ok, noted. Changes will be commited.

**Review by alexorlenko — APPROVED** [review]

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/GlobalExceptionMapper.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+
+@Provider
+public class GlobalExceptionMapper implements ExceptionMapper<Exception> {
+
+
+    @Override
+    public Response toResponse(Exception exception) {
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
```
Comment (kh0ma) [inline home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/GlobalExceptionMapper.java]: Is this mapping following our contract in open API definition?

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/exception/BaseApiException.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.softserveinc.ita.homeproject.homeservice.exception;
+
+public abstract class BaseApiException extends RuntimeException {
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/exception/BaseApiException.java]: The service layer should know nothing about presentation so we cannot use anything from API here. From the package I see that this exception belongs to service layer, please remove API mentioning from the naming:
```
BaseHomeException
  NotFoundHomeException
  AlreadyExsistHomeException
```

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/BaseApiExceptionMapper.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Abstract class for handling the mapping to the response
+ * Changeable data is in abstract method.
+ *
+ * @param <T> type of incoming exception
+ *
+ * @author Mykyta Morar
+ */
+public abstract class BaseApiExceptionMapper<T extends BaseApiException>  implements ExceptionMapper<T> {
```
Comment (kh0ma) [inline home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/BaseApiExceptionMapper.java]: Please consider the following structure:
```java
public abstract class BaseExceptionMapper<T extends Throwable> implements ExceptionMapper<T> {

    @Override
    public Response toResponse(T exception) {
    	Response.Status status = getStatus();
        return Response.status(status)
                .entity(new ApiError()
                        .responseCode(status.getStatusCode())
                        .errorMessage(extractMessage(exception))
                .build();
    }

    protected abstract Response.Status getStatus();
        
    protected abstract String extractMessage(T exception);

}

public class GeneralExceptionMapper extends BaseExceptionMapper<Trowable> {

    protected Response.Status getStatus() {
    	return INTERNAL_SERVER_ERROR;
    }
        
    protected String extractMessage(Trowable exception) {
    	return "Unknown error. Please contact support.";
    }

}

public abstract class BaseHomeExceptionMapper<T extends BaseHomeException> extends BaseExceptionMapper<T> {
        
    protected String extractMessage(T exception) {
    	return exception.getMessage();
    }

}


public class NotFoundHomeExceptionMapper extends BaseHomeExceptionMapper<NotFoundHomeException> {

    protected Response.Status getStatus() {
    	return ...; //404
    }

}


public class AlreadExistHomeExceptionMapper extends BaseHomeExceptionMapper<AlreadExistHomeException> {

    protected Response.Status getStatus() {
    	return ...; //400
    }

}
```

**Review by kh0ma — CHANGES_REQUESTED** [review]

**Review by kh0ma — APPROVED** [review]

**Review by PoLytvy — APPROVED** [review]

**Review by Romanov-Niko — APPROVED** [review]

**Pull request merged**